### PR TITLE
Microbenchmarks

### PR DESCRIPTION
**Pull request by johndaniels**

### CODE REVIEW

File: test/cppsuite/configs/session_microbenchmarks_default.txt [hunk 1]
```diff
@@ -0,0 +1,19 @@
+# Configuration for session_microbenchmarks.
+duration_seconds=5,
+cache_size_mb=250,
```
Comment (luke-pearson) [inline test/cppsuite/configs/session_microbenchmarks_default.txt]: Add: 
cache_size_mb=2000,
validate=false,
in_memory=true,
sweep_interval=1000,

File: test/cppsuite/tests/session_microbenchmarks.cpp [hunk 2]
```diff
@@ -0,0 +1,230 @@
...
+ */
+
+#include "src/util/instruction_counter.h"
+#include "src/common/constants.h"
+#include "src/common/logger.h"
+#include "src/main/test.h"
+
+namespace test_harness {
+/* Defines what data is written to the tracking table for use in custom validation. */
+class operation_tracker_session_microbenchmarks : public operation_tracker {
```
Comment (luke-pearson) [inline test/cppsuite/tests/session_microbenchmarks.cpp]: The whole operation_tracker class can be removed.

File: test/cppsuite/tests/session_microbenchmarks.cpp [hunk 3]
```diff
@@ -0,0 +1,230 @@
...
+    {
+        /* You can remove the call to the base class to fully customize your test. */
+        test::run();
+    }
+
+    void
+    background_compact_operation(thread_worker *) override final
+    {
+        logger::log_msg(LOG_WARN, "background_compact_operation: nothing done");
+    }
```
Comment (luke-pearson) [inline test/cppsuite/tests/session_microbenchmarks.cpp]: Remove

File: test/cppsuite/tests/session_microbenchmarks.cpp [hunk 4]
```diff
@@ -0,0 +1,230 @@
...
+
+    void
+    checkpoint_operation(thread_worker *) override final
+    {
+        logger::log_msg(LOG_WARN, "checkpoint_operation: nothing done");
```
Comment (luke-pearson) [inline test/cppsuite/tests/session_microbenchmarks.cpp]: Update to be something like. "checkpoint_operation: not done as this is a performance test".

File: test/cppsuite/tests/session_microbenchmarks.cpp [hunk 5]
```diff
@@ -0,0 +1,230 @@
...
+          "timestamp_transaction_uint_instructions", test::_args.test_name);
+        std::string cursor_uri = tc->db.get_collection(0).name;
+
+        /* Get the collection to work on. */
+        testutil_assert(tc->collection_count == 1);
+
+        scoped_session &session = tc->session;
+
+        begin_transaction_counter.track([&wt_session, &session]() -> int {
+            session->begin_transaction(session.get(), NULL);
```
Comment (luke-pearson) [inline test/cppsuite/tests/session_microbenchmarks.cpp]: I think we want to assert outside the track call that the begin returned success. Same for the other track calls.

File: test/cppsuite/tests/session_microbenchmarks.cpp [hunk 6]
```diff
@@ -0,0 +1,230 @@
...
+        make_insert(tc, "1");
+
+        commit_transaction_counter.track([&wt_session, &session]() -> int {
+            session->commit_transaction(session.get(), NULL);
+            return 0;
+        });
+
+        session->begin_transaction(session.get(), NULL);
+
+        make_insert(tc, "2");
```
Comment (luke-pearson) [inline test/cppsuite/tests/session_microbenchmarks.cpp]: To reduce the additional work done in this rollback we don't need the insert here.

File: test/cppsuite/tests/session_microbenchmarks.cpp [hunk 7]
```diff
@@ -0,0 +1,230 @@
...
+
+        rollback_trancation_counter.track([&wt_session, &session]() -> int {
+            session->rollback_transaction(session.get(), NULL);
+            return 0;
+        });
+
+        std::string prepare_timestamp = tc->tsm->decimal_to_hex(tc->tsm->get_next_ts());
```
Comment (luke-pearson) [inline test/cppsuite/tests/session_microbenchmarks.cpp]: We could use a single byte timestamp here but I don't know how much that would affect the number of instructions.

File: test/cppsuite/tests/session_microbenchmarks.cpp [hunk 8]
```diff
@@ -0,0 +1,230 @@
...
+
+        commit_after_prepare_transaction_counter.track([&session, &commit_timestamp]() -> int {
+            session->commit_transaction(session.get(),
+              ("commit_timestamp=" + commit_timestamp + ",durable_timestamp=" + commit_timestamp)
+                .c_str());
+            return 0;
+        });
+
+        session->begin_transaction(session.get(), NULL);
+        make_insert(tc, "4");
```
Comment (luke-pearson) [inline test/cppsuite/tests/session_microbenchmarks.cpp]: Can remove this insert.

File: test/cppsuite/tests/session_microbenchmarks.cpp [hunk 9]
```diff
@@ -0,0 +1,230 @@
...
+    update_operation(thread_worker *) override final
+    {
+        logger::log_msg(LOG_WARN, "update_operation: nothing done");
+    }
+
+    void
+    validate(bool, const std::string &, const std::string &, database &) override final
+    {
+        logger::log_msg(LOG_WARN, "validate: nothing done");
+    }
```
Comment (luke-pearson) [inline test/cppsuite/tests/session_microbenchmarks.cpp]: Remove.

File: test/cppsuite/tests/session_microbenchmarks.cpp (hunk 7 above)
Comment (johndaniels) [reply test/cppsuite/tests/session_microbenchmarks.cpp]: When I tried that it said the timestamp wasn't big enough (that it needed to be larger than all the previous timestamps, I can track down the error if that helps).

File: test/cppsuite/tests/bounded_cursor_perf.cpp [hunk 10]
```diff
@@ -129,8 +130,9 @@ class bounded_cursor_perf : public test {
             }
 
             /* Advance the cursors one position. */
-            auto range_ret_next = bounded_next.track([&next_range_cursor]() -> int {
-                return next_range_cursor->next(next_range_cursor.get());
+            int test_closure;
```
Comment (luke-pearson) [inline test/cppsuite/tests/bounded_cursor_perf.cpp]: I think this was a test of closure semantics, which we can remove?

File: test/cppsuite/tests/session_microbenchmarks.cpp [hunk 11]
```diff
@@ -0,0 +1,171 @@
...
+        instruction_counter timestamp_transaction_uint_counter(
+          "timestamp_transaction_uint_instructions", test::_args.test_name);
+        std::string cursor_uri = tc->db.get_collection(0).name;
+
+        /* Get the collection to work on. */
+        testutil_assert(tc->collection_count == 1);
+
+        scoped_session &session = tc->session;
+
+        int result = begin_transaction_counter.track([&wt_session, &session]() -> int {
```
Comment (luke-pearson) [inline test/cppsuite/tests/session_microbenchmarks.cpp]: Seeing as we're using the c++ scoped session we can capture that only and remove the `wt_sessions` var entirely.

File: test/cppsuite/tests/session_microbenchmarks.cpp (hunk 7 above)
Comment (luke-pearson) [reply test/cppsuite/tests/session_microbenchmarks.cpp]: Oh of course because the populate stage is using MongoDB style timestamps.

**Pull request merged**